lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,608 @@
+    userContext: UserContext;
+}) => string; // should return the domain of the origin
+
+export type TypeNormalisedInputRelyingPartyName = (input: {
+    tenantId: string;
+    userContext: UserContext;
+}) => Promise<string>; // should return the app name
+
+export type TypeNormalisedInputGetOrigin = (input: {
+    tenantId: string;
+    request: BaseRequest;
+    userContext: UserContext;
+}) => Promise<string>; // should return the app name
+
+export type TypeInput = {
+    emailDelivery?: EmailDeliveryTypeInput<TypeWebauthnEmailDeliveryInput>;
+    relyingPartyId?: TypeInputRelyingPartyId;
+    relyingPartyName?: TypeInputRelyingPartyName;
+    getOrigin?: TypeInputGetOrigin;
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeInputRelyingPartyId =
+    | string
+    | ((input: { tenantId: string; request: BaseRequest | undefined; userContext: UserContext }) => Promise<string>);
+
+export type TypeInputRelyingPartyName =
+    | string
+    | ((input: { tenantId: string; userContext: UserContext }) => Promise<string>);
+
+export type TypeInputGetOrigin = (input: {
+    tenantId: string;
+    request: BaseRequest;
+    userContext: UserContext;
+}) => Promise<string>;
+
+export type RecipeInterface = {
+    // should have a way to access the user email: passed as a param, through session, or using recoverAccountToken
+    // it should have at least one of those 3 options
+    registerOptions(
+        input: {
+            relyingPartyId: string;
+            relyingPartyName: string;
+            origin: string;
+            requireResidentKey: boolean | undefined; // should default to false in order to allow multiple authenticators to be used; see https://auth0.com/blog/a-look-at-webauthn-resident-credentials/
+            // default to 'required' in order store the private key locally on the device and not on the server
+            residentKey: "required" | "preferred" | "discouraged" | undefined;
+            // default to 'preferred' in order to verify the user (biometrics, pin, etc) based on the device preferences
+            userVerification: "required" | "preferred" | "discouraged" | undefined;
+            // default to 'none' in order to allow any authenticator and not verify attestation
+            attestation: "none" | "indirect" | "direct" | "enterprise" | undefined;
+            // default to 5 seconds
+            timeout: number | undefined;
+            tenantId: string;
+            userContext: UserContext;
+        } & (
+            | {
+                  recoverAccountToken: string;
+              }
+            | {
+                  email: string;
+              }
+            | {
+                  session: SessionContainerInterface;
```

Comment:
when is only a session needed? And how will we get the email from the session? What if the currently logging in user has no email, or what if the currently logged in user has multiple emails (cause they are a primary user with multiple emails)

Reply:
If there is no email, it will throw. I've added the error. As for when there are multiple emails i was thinking of only using the email from the webauthn login method. Basically allowing register passkey to be used if the user is logged in with webauthn. If there is no session, they should pass one of the other options...

What do you think ?

Reply:
I dont think its clear when a session will be an input.. it could mean that the user wants to add a passkey to an existing logged in account. In that case, there could be already a webauthn login method, or there may not be any at all.. so we cant just throw in case an email doesn't exist, and we can't just assume webauthn login method is always there. We need clarity on the flow here! Please think about this case carefully.

Reply:
i think i'm missing context then. doesn't the existence of webauthn in the loginMethods mean that the webauth loging method exists ? Maybe we can chat

Reply:
we decided to skip using the session